### PR TITLE
Feature/repo 5181 implement child assoc events

### PR DESCRIPTION
**Pull request by SaraAspery**

### CODE REVIEW

File: src/main/java/org/alfresco/repo/event2/PeerAssociationEventConsolidator.java [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+ *
+ * @author Sara Aspery
+ */
+public class PeerAssociationEventConsolidator implements PeerAssociationEventSupportedPolicies
+{
+    private final Deque<EventType> eventTypes;
+
+    private AssociationRef associationRef;
+    PeerAssociationResource resource;
+    QNameHelper helper;
```
[the remark this answers is not in the document]
Comment (jamalkm) [reply src/main/java/org/alfresco/repo/event2/PeerAssociationEventConsolidator.java:48]: Same as the other consolidator, `associationRef` and `helper` can be final.

File: src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+public class EventConsolidatorUnitTest extends TestCase
+{
+    private NodeResourceHelper nodeResourceHelper = mock(NodeResourceHelper.class);
+    
+    @Test
+    public void testGetMappedAspectsBeforeRemovedAndAddedEmpty()
+    {
+        EventConsolidator eventConsolidator = new EventConsolidator(nodeResourceHelper);
+        
+        Set<String> currentApects = new HashSet<>();
```
Comment (jamalkm) [inline src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java:47]: typo: currentAspects

File: src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+        assertEquals(0, mappedAspectsBefore.size());
+    }
+
+    @Test
+    public void testGetMappedAspectsBefore_AspectRemoved()
+    {
+        EventConsolidator eventConsolidator = new EventConsolidator(nodeResourceHelper);
+        eventConsolidator.addAspect(ContentModel.ASSOC_CONTAINS);
+
+        Set<String> currentApects = new HashSet<>();
```
Comment (jamalkm) [inline src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java:62]: typo: currentAspects

File: src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+        System.out.println(mappedAspectsBefore);
+    }
+
+    @Test
+    public void testGetMappedAspectsBefore_AspectAdded()
+    {
+        EventConsolidator eventConsolidator = new EventConsolidator(nodeResourceHelper);
+        eventConsolidator.addAspect(ContentModel.ASSOC_CONTAINS);
+
+        Set<String> currentApects = new HashSet<>();
```
Comment (jamalkm) [inline src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java:85]: Same typo. I'm not going to add comment for others, so please make sure all `currentApects` variables are changed to `currentAspects`

File: src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+        Set<String> added = new HashSet<>();
+        removed.add("cm:contains");
+
+        when(nodeResourceHelper.mapToNodeAspects(eventConsolidator.getAspectsRemoved())).thenReturn(removed);
+        when(nodeResourceHelper.mapToNodeAspects(eventConsolidator.getAspectsAdded())).thenReturn(added);
+
+        Set<String> mappedAspectsBefore = eventConsolidator.getMappedAspectsBefore(currentApects);
+
+        assertEquals(3, mappedAspectsBefore.size());
+        System.out.println(mappedAspectsBefore);
```
Comment (jamalkm) [inline src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java:76]: leftover? System.out

File: src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+        Set<String> added = new HashSet<>();
+        added.add("cm:auditable");
+
+        when(nodeResourceHelper.mapToNodeAspects(eventConsolidator.getAspectsRemoved())).thenReturn(removed);
+        when(nodeResourceHelper.mapToNodeAspects(eventConsolidator.getAspectsAdded())).thenReturn(added);
+
+        Set<String> mappedAspectsBefore = eventConsolidator.getMappedAspectsBefore(currentApects);
+
+        assertEquals(1, mappedAspectsBefore.size());
+        System.out.println(mappedAspectsBefore);
```
Comment (jamalkm) [inline src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java:99]: System.out

File: src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+        Set<String> added = new HashSet<>();
+        added.add("cm:contains");
+        
+        when(nodeResourceHelper.mapToNodeAspects(eventConsolidator.getAspectsRemoved())).thenReturn(removed);
+        when(nodeResourceHelper.mapToNodeAspects(eventConsolidator.getAspectsAdded())).thenReturn(added);
+        
+        Set<String> mappedAspectsBefore = eventConsolidator.getMappedAspectsBefore(currentApects);
+        
+        assertEquals(2, mappedAspectsBefore.size());
+        System.out.println(mappedAspectsBefore);
```
Comment (jamalkm) [inline src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java:124]: System.out

File: src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+        
+        Set<String> mappedAspectsBefore = eventConsolidator.getMappedAspectsBefore(currentApects);
+        
+        System.out.println("Before: " + mappedAspectsBefore);
+        assertEquals(2, mappedAspectsBefore.size());
+    }
+
+
+    @Test
+    public void testGetMappedAspectsBefore_AspectRemovedTwiceAddeddOnce()
```
Comment (jamalkm) [inline src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java:183]: typo: method name `...TwiceAddeddOnce()` -> `TwiceAddedOnce()`

File: src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+        eventConsolidator.removeAspect(ContentModel.ASSOC_CONTAINS);
+
+        Set<String> currentApects = new HashSet<>();
+        currentApects.add("cm:geographic");
+        currentApects.add("cm:auditable");
+        currentApects.add("cm:contains");
+
+        Set<String> removed = new HashSet<>();
+        removed.add("cm:contains");
+        removed.add("cm:contains");
```
Comment (jamalkm) [inline src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java:199]: Duplicate set element!

File: src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+import static org.mockito.Mockito.when;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import junit.framework.TestCase;
+import org.alfresco.model.ContentModel;
+import org.junit.Test;
+
+public class EventConsolidatorUnitTest extends TestCase
```
Comment (jamalkm) [inline src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java:38]: Mix of JUnit 4 (`@Test`) and JUnit 3 (`TestCase`).
There is no need to extend `TestCase`.
For `assertEquals` and etc. use static import of  `org.junit.<Your-Required-Method>`

File: src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java [hunk 11]
```diff
@@ -0,0 +1,303 @@
...
+        EventConsolidator eventConsolidator = new EventConsolidator(nodeResourceHelper);
+        eventConsolidator.addAspect(ContentModel.ASSOC_CONTAINS);
+        eventConsolidator.removeAspect(ContentModel.ASSOC_CONTAINS);
+
+        assertEquals(0, eventConsolidator.getAspectsAdded().size());
+        assertEquals(0, eventConsolidator.getAspectsRemoved().size());
+    }
+
+    @Test
+    public void testRemvoeAspectAddAspect()
```
Comment (jamalkm) [inline src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java:268]: typo: `testRemvoe...` -> `testRemove...`

File: src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java [hunk 12]
```diff
@@ -0,0 +1,303 @@
...
+        Set<String> removed = new HashSet<>();
+        Set<String> added = new HashSet<>();
+        added.add("cm:contains");
+        
+        when(nodeResourceHelper.mapToNodeAspects(eventConsolidator.getAspectsRemoved())).thenReturn(removed);
+        when(nodeResourceHelper.mapToNodeAspects(eventConsolidator.getAspectsAdded())).thenReturn(added);
+        
+        Set<String> mappedAspectsBefore = eventConsolidator.getMappedAspectsBefore(currentApects);
+        
+        System.out.println("Before: " + mappedAspectsBefore);
```
Comment (jamalkm) [inline src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java:177]: System.out

File: src/test/java/org/alfresco/repo/event2/UpdateRepoEventIT.java [hunk 13]
```diff
@@ -820,4 +827,62 @@ public void testCreateAndMoveFileInTheSameTransaction()
...
+    {
+        final NodeRef nodeRef = createNode(ContentModel.TYPE_CONTENT);
+        NodeResource resource = getNodeResource(1);
+        final Set<String> originalAspects = resource.getAspectNames();
+        assertNotNull(originalAspects);
+
+
+        retryingTransactionHelper.doInTransaction(() -> {
+            // Add cm:geographic aspect with default value
+            nodeService.addAspect(nodeRef, ContentModel.ASPECT_GEOGRAPHIC, null);
```
Comment (jamalkm) [inline src/test/java/org/alfresco/repo/event2/UpdateRepoEventIT.java:842]: I think this test will fail if we change it to also include property. E.g. `nodeService.addAspect(nodeRef, ContentModel.ASPECT_GEOGRAPHIC, Map.of(ContentModel.PROP_LATITUDE, 52.53122d));`

Anyway I just left comment here to remind myself later, it is not a blocker for this PR.

File: src/main/java/org/alfresco/repo/event2/ChildAssociationEventConsolidator.java [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+ * @author Chris Shields
+ * @author Sara Aspery
+ */
+public class ChildAssociationEventConsolidator implements ChildAssociationEventSupportedPolicies
+{
+    private final Deque<EventType> eventTypes;
+
+    private ChildAssociationRef childAssociationRef;
+    private ChildAssociationResource resource;
+    QNameHelper helper;
```
[the remark this answers is not in the document]
Comment (cmshields) [reply src/main/java/org/alfresco/repo/event2/ChildAssociationEventConsolidator.java]: Done

File: src/main/java/org/alfresco/repo/event2/ChildAssociationEventConsolidator.java [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+
+    public ChildAssociationEventConsolidator(ChildAssociationRef childAssociationRef, QNameHelper helper)
+    {
+        this.eventTypes = new ArrayDeque<>();
+        this.childAssociationRef = childAssociationRef;
+        this.helper = helper;
+        resource = buildChildAssociationResource(this.childAssociationRef);
```
[the remark this answers is not in the document]
Comment (cmshields) [reply src/main/java/org/alfresco/repo/event2/ChildAssociationEventConsolidator.java]: Done

File: src/main/java/org/alfresco/repo/event2/EventConsolidator.java [hunk 16]
```diff
@@ -348,7 +387,15 @@ private NodeResource buildNodeResourceBeforeDelta(NodeResource after)
...
     {
-        return !(aspectsRemoved.isEmpty() && aspectsAdded.isEmpty());
+        if (aspectsRemoved.isEmpty() && aspectsAdded.isEmpty())
+        {
+            return false;
+        }
+        else if (aspectsRemoved.size() == aspectsAdded.size() && aspectsRemoved.containsAll(aspectsAdded))
+        {
+            return false;
+        }
```
[the remark this answers is not in the document]
Comment (cmshields) [reply src/main/java/org/alfresco/repo/event2/EventConsolidator.java]: Done

File: src/main/java/org/alfresco/repo/event2/EventGenerator.java [hunk 17]
```diff
@@ -296,18 +394,77 @@ private void sendEvent(NodeRef nodeRef, EventConsolidator consolidator)
...
+                        + ((childAssocType == null) ? "Unknown' " : childAssocType.toPrefixString())
+                        + "' created by: " + user);
+            }
+            return;
+        }
+
+        logAndSendEvent(event, consolidator.getEventTypes());
+    }
+
+    private void sendEvent(AssociationRef peerAssociationRef, PeerAssociationEventConsolidator consolidator)
```
[the remark this answers is not in the document]
Comment (cmshields) [reply src/main/java/org/alfresco/repo/event2/EventGenerator.java]: Done

File: src/main/java/org/alfresco/repo/event2/EventGenerator.java [hunk 18]
```diff
@@ -322,25 +479,91 @@ protected void onShutdown(ApplicationEvent applicationEvent)
...
             }
         }
     }
+
+
+    private class Consolidators
+    {
+        Map<NodeRef, EventConsolidator> nodes;
+        Map<ChildAssociationRef, ChildAssociationEventConsolidator> childAssocs;
+        Map<AssociationRef, PeerAssociationEventConsolidator> peerAssocs;
```
[the remark this answers is not in the document]
Comment (cmshields) [reply src/main/java/org/alfresco/repo/event2/EventGenerator.java]: Done

File: src/main/java/org/alfresco/repo/event2/EventType.java [hunk 19]
```diff
@@ -32,21 +32,23 @@
  */
 public enum EventType
 {
-    NODE_CREATED("Created"), NODE_UPDATED("Updated"), NODE_DELETED("Deleted"), NODE_DOWNLOADED("Downloaded");
+    NODE_CREATED("Created", ContextType.NODE), NODE_UPDATED("Updated", ContextType.NODE), NODE_DELETED("Deleted", ContextType.NODE), NODE_DOWNLOADED("Downloaded", ContextType.NODE),
+    CHILD_ASSOC_CREATED("Created", ContextType.CHILD_ASSOC), CHILD_ASSOC_DELETED("Deleted", ContextType.CHILD_ASSOC),
+    PEER_ASSOC_CREATED("Created", ContextType.PEER_ASSOC), PEER_ASSOC_DELETED("Deleted", ContextType.PEER_ASSOC);
```
[the remark this answers is not in the document]
Comment (cmshields) [reply src/main/java/org/alfresco/repo/event2/EventType.java]: Done

File: src/main/java/org/alfresco/repo/event2/PeerAssociationEventConsolidator.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (cmshields) [reply src/main/java/org/alfresco/repo/event2/PeerAssociationEventConsolidator.java]: Done

File: src/main/java/org/alfresco/repo/event2/EventGenerator.java [hunk 20]
```diff
@@ -322,25 +479,91 @@ protected void onShutdown(ApplicationEvent applicationEvent)
...
             catch (Exception e)
             {
                 // Must consume the exception to protect other TransactionListeners
                 LOGGER.error("Unexpected error while sending repository events", e);
             }
         }
     }
+
+
+    private class Consolidators
```
[the remark this answers is not in the document]
Comment (cmshields) [reply src/main/java/org/alfresco/repo/event2/EventGenerator.java]: Done. Inner class now static.

File: src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java (hunk 2 above)
Comment (cmshields) [reply src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java]: Done

File: src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java (hunk 3 above)
Comment (cmshields) [reply src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java]: Done

File: src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java (hunk 4 above)
Comment (cmshields) [reply src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java]: Done

File: src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java (hunk 5 above)
Comment (cmshields) [reply src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java]: Done

File: src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java (hunk 6 above)
Comment (cmshields) [reply src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java]: Done

File: src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java (hunk 7 above)
Comment (cmshields) [reply src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java]: Done

File: src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java (hunk 8 above)
Comment (cmshields) [reply src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java]: Done

File: src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java (hunk 9 above)
Comment (cmshields) [reply src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java]: Done

File: src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java (hunk 10 above)
Comment (cmshields) [reply src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java]: Done

File: src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java (hunk 11 above)
Comment (cmshields) [reply src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java]: Done

File: src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java (hunk 12 above)
Comment (cmshields) [reply src/test/java/org/alfresco/repo/event2/EventConsolidatorUnitTest.java]: Done

**Pull request merged**